xapian-applications/omega/handler_libarchive.cc

```diff
@@ -31,92 +34,262 @@
 
 using namespace std;
 
-bool
-extract(const string& filename,
-	string& dump,
-	string& title,
-	string& keywords,
-	string& author,
-	string& pages,
-	string& error)
+static void
+parse_metadata(string& s_meta,
+	       string& title,
+	       string& keywords,
+	       string& author)
 {
-    try {
-	const char* file = filename.c_str();
-	struct archive* archive_obj = archive_read_new();
-	archive_read_support_format_zip(archive_obj);
-	// Block size will be determined by libarchive automatically for
-	// regular files. Specified block size will only be used for tape drives
-	// 10240 is chosen as default size (20 records - 512 bytes each)
-	int status_code = archive_read_open_filename(archive_obj, file,
-						     DEFAULT_BLOCK_SIZE);
+    MetaXmlParser metaxmlparser;
+    metaxmlparser.parse(s_meta);
+    title = metaxmlparser.title;
+    keywords = metaxmlparser.keywords;
+    author = metaxmlparser.author;
+}
 
-	if (status_code != ARCHIVE_OK) {
-	    error = "Libarchive failed to open the file " + filename;
-	    return false;
+static bool
+extract_odf(struct archive* archive_obj,
+	    string& dump,
+	    string& title,
+	    string& keywords,
+	    string& author,
+	    string& error)
+{
+    string s_content, s_styles;
+    struct archive_entry* entry;
+    size_t total;
+    ssize_t size;
+    string s_meta;
+
+    while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	string pathname = archive_entry_pathname(entry);
+	if (pathname == "content.xml") {
+	    total = archive_entry_size(entry);
+	    s_content.resize(total);
+	    size = archive_read_data(archive_obj, &s_content[0], total);
+
+	    if (size <= 0) {
+		error = "Libarchive was not able to extract data from "
+			"content.xml";
+		return false;
+	    }
+	    s_content.resize(size);
+	} else if (pathname == "styles.xml") {
+	    total = archive_entry_size(entry);
+	    s_styles.resize(total);
+	    size = archive_read_data(archive_obj, &s_styles[0], total);
+
+	    if (size <= 0) {
+		error = "Libarchive was not able to extract data from "
+			"styles.xml";
+		return false;
+	    }
+	    s_styles.resize(size);
+	} else if (pathname == "meta.xml") {
+		total = archive_entry_size(entry);
+		s_meta.resize(total);
+		size = archive_read_data(archive_obj, &s_meta[0], total);
+
+		if (size > 0) {
+		    // indexing file even if this fails
+		    parse_metadata(s_meta, title, keywords, author);
+		}
+	    s_meta.resize(size);
```

Comment:
Here I think there is a problem.
The type of `size` is `ssize_t`. In short, `ssize_t` is similar to `size_t`(unsigned) but it can store -1 to notify an error.
It seems that if `archive_read_data` fails, size will hold -1 and an exception is going to be raised if you call resize with negative values.

Reply:
Indeed - `resize(-1)` on a string will cast the -1 to the largest value of the unsigned `std::string::size_type`, which will (hopefully) be larger than you can actually resize a string to.

Also, to be useful this resize needs to happen *before* we pass `s_meta` to `parse_metadata()` (though you'd usually expect the number of bytes read to match the size reported for the entry in `total`, so we probably get away with this most of the time).

Reply:
Thanks for pointing this out. I have changed this. Now resize() is called before it is sent to parse_metadata().